integration/metrics_test.go

```diff
@@ -192,3 +201,189 @@ disable_verification = %v
+				indexDigest := optimizeImageWithOpts(sh, image, 1<<22, 0)
+				return buildIndexByManipulatingZtocData(t, sh, indexDigest)
+			},
+			expectFuseOperationFailure: true,
+		},
+	}
+
+	// checkFuseOperationFailureMetrics checks if output from metrics endpoint includes
+	// any fuse operation failure metrics
+	checkFuseOperationFailureMetrics := func(t *testing.T, output string, expectFuseOperationFailure bool) {
+		metricCountSum := 0
+
+		lines := strings.Split(output, "\n")
+		for _, line := range lines {
+			// filter out non-fuse and fuse_mount_failure_count metrics
+			if !strings.Contains(line, "fuse") || strings.Contains(line, commonmetrics.FuseMountFailureCount) {
+				continue
+			}
+
+			parts := strings.Split(line, " ")
+			if metricCount, err := strconv.Atoi(parts[len(parts)-1]); err == nil && metricCount != 0 {
+				if !expectFuseOperationFailure {
+					t.Fatalf("unexpected fuse operation failure: %s", line)
+				}
+				t.Logf("fuse operation failure metric: %s", line)
+				metricCountSum += metricCount
+			}
+		}
+
+		if (metricCountSum != 0) != expectFuseOperationFailure {
+			t.Fatalf("invalid fuse operation failure metrics. total operation failure count: %d, expect fuse operation failure: %t",
+				metricCountSum, expectFuseOperationFailure)
+		}
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			rebootContainerd(t, sh, "", "")
+
+			imgInfo := dockerhub(tc.image)
+			sh.X("ctr", "i", "pull", imgInfo.ref)
+			indexDigest := tc.indexDigestFn(t, sh, imgInfo)
+
+			sh.X("soci", "image", "rpull", "--soci-index-digest", indexDigest, imgInfo.ref)
+			// this command may fail due to fuse operation failure, refresh the shell
+			sh.XLog("ctr", "run", "--rm", "--snapshotter=soci", imgInfo.ref, "test", "echo", "hi")
+			sh = sh.Refresh()
+
+			curlOutput := string(sh.O("curl", "localhost:6060/metrics"))
+			checkFuseOperationFailureMetrics(t, curlOutput, tc.expectFuseOperationFailure)
+		})
+	}
+}
+
+// buildIndexByManipulatingZtocData produces a new soci index by manipulating
+// the ztocs of an existing index specified by `indexDigest`.
+//
+// The new index (and ztocs) are stored separately and the original index keeps unchanged.
+// The manipulated ztocs are (de)serializable but have meaningless ztoc data. This helps
+// test soci behaviors when ztocs have valid format but wrong/corrupted data.
+func buildIndexByManipulatingZtocData(t *testing.T, sh *shell.Shell, indexDigest string) string {
+	index := sociIndexFromDigest(t, sh, indexDigest)
+
+	var ztocDescs []ocispec.Descriptor
+	for _, blob := range index.Blobs {
+		ztocDigest := blob.Digest.String()
+		blobContent := fetchContentFromPath(sh, blobStorePath+"/"+trimSha256Prefix(ztocDigest))
+		zt, err := ztoc.Unmarshal(bytes.NewReader(blobContent))
+		if err != nil {
+			t.Fatalf("invalid ztoc %s from soci index %s: %v", ztocDigest, indexDigest, err)
+		}
+
+		// manipulate the ztoc
+		for i, md := range zt.TOC.Metadata {
```

Comment:
I think we can have another test case where fs metadata is correct but dictionaries are corrupted.

Reply:
do you mean this `CompressionInfo`? what's the expected fuse operation failures by manipulating these 3 fields?

https://github.com/awslabs/soci-snapshotter/blob/e83d75262b736cff39f44d7e66ab7422f2837792/ztoc/ztoc.go#L60-L62

Reply:
Yes. Incorrect `SpanDigests` would mean that compressed data will not be available locally. Incorrect `Checkpoints` would mean that decompression will fail.

Reply:
I think we can add it in a follow up PR. Here I refactored the `buildIndexByManipulatingZtocData` to accept a `manipulator` func, so it's easy to try different manipulations.

I found even valid index/ztoc produces some fuse failures, so we need to first figure out _what ztoc manipulation leads to which fuse failure_, and then check the specific fuse failure metric.